auth/src/main/java/com/firebase/ui/auth/util/smartlock/SignInDelegate.java

```diff
@@ -0,0 +1,427 @@
+                            // In this case the credential saved in SmartLock was not
+                            // a valid credential, we should delete it from SmartLock
+                            // before continuing.
+                            deleteCredentialAndRedirect();
+                        } else {
+                            startAuthMethodChoice();
+                        }
+                    }
+                });
+    }
+
+    /**
+     * Delete the last credential retrieved from SmartLock and then redirect to the
+     * auth method choice flow.
+     */
+    private void deleteCredentialAndRedirect() {
+        if (mCredential == null) {
+            Log.w(TAG, "deleteCredentialAndRedirect: null credential");
+            startAuthMethodChoice();
+            return;
+        }
+
+        CredentialsApiHelper credentialsApiHelper = CredentialsApiHelper.getInstance(getActivity());
+        credentialsApiHelper.delete(mCredential)
+                .addOnCompleteListener(new OnCompleteListener<Status>() {
+                    @Override
+                    public void onComplete(@NonNull Task<Status> task) {
+                        if (!task.isSuccessful()) {
+                            Log.w(TAG, "deleteCredential:failure", task.getException());
+                        }
+                        startAuthMethodChoice();
+                    }
+                });
+    }
+
+    private void redirectToIdpSignIn(String email, String accountType) {
+        Intent nextIntent;
+        switch (accountType) {
+            case IdentityProviders.GOOGLE:
+                nextIntent = IdpSignInContainerActivity.createIntent(
+                        getContext(),
+                        mHelper.getFlowParams(),
+                        GoogleAuthProvider.PROVIDER_ID,
+                        email);
+                break;
+            case IdentityProviders.FACEBOOK:
+                nextIntent = IdpSignInContainerActivity.createIntent(
+                        getContext(),
+                        mHelper.getFlowParams(),
+                        FacebookAuthProvider.PROVIDER_ID,
+                        email);
+                break;
+            case IdentityProviders.TWITTER:
+                nextIntent = IdpSignInContainerActivity.createIntent(
+                        getContext(),
+                        mHelper.getFlowParams(),
+                        TwitterAuthProvider.PROVIDER_ID,
+                        email);
+                break;
+            default:
+                Log.w(TAG, "unknown provider: " + accountType);
+                nextIntent = AuthMethodPickerActivity.createIntent(
+                        getContext(),
+                        mHelper.getFlowParams());
+        }
+        startActivityForResult(nextIntent, RC_IDP_SIGNIN);
+    }
+
+    private void finish(int resultCode, Intent data) {
+        if (mGoogleApiClient != null) {
+            mGoogleApiClient.disconnect();
```

Comment:
Yup, `onDestroy()` is called if the parent activity dies so I added some cleanup to `SmartLock` because I noticed that we weren't doing anything for `SaveSmartLock` either.